In [5]:
import pandas as pd
import numpy as np
import os
import gensim
import nltk

In [6]:
nltk.download('all')

[nltk_data] Downloading collection 'all'
[nltk_data]    | 
[nltk_data]    | Downloading package abc to
[nltk_data]    |     C:\Users\dhiya\AppData\Roaming\nltk_data...
[nltk_data]    |   Unzipping corpora\abc.zip.
[nltk_data]    | Downloading package alpino to
[nltk_data]    |     C:\Users\dhiya\AppData\Roaming\nltk_data...
[nltk_data]    |   Unzipping corpora\alpino.zip.
[nltk_data]    | Downloading package averaged_perceptron_tagger to
[nltk_data]    |     C:\Users\dhiya\AppData\Roaming\nltk_data...
[nltk_data]    |   Unzipping taggers\averaged_perceptron_tagger.zip.
[nltk_data]    | Downloading package averaged_perceptron_tagger_eng to
[nltk_data]    |     C:\Users\dhiya\AppData\Roaming\nltk_data...
[nltk_data]    |   Unzipping
[nltk_data]    |       taggers\averaged_perceptron_tagger_eng.zip.
[nltk_data]    | Downloading package averaged_perceptron_tagger_ru to
[nltk_data]    |     C:\Users\dhiya\AppData\Roaming\nltk_data...
[nltk_data]    |   Unzipping
[nltk_data]    |       tagge

True

In [7]:
from nltk import sent_tokenize, word_tokenize
from gensim.utils import simple_preprocess



In [10]:
for file in os.listdir('data'):
    print(file)

story1.txt
story2.txt


In [17]:
from pathlib import Path


In [43]:
current_dir = Path.cwd()

#base_path = Path(__file__).resolve().parent.parent
data_path = current_dir/"data"

print(current_dir)
print(data_path)


d:\krishnaik_AI\FULL-STACK-GENAI-BOOTCAMP-1.0\Class-04-12-APR
d:\krishnaik_AI\FULL-STACK-GENAI-BOOTCAMP-1.0\Class-04-12-APR\data


In [32]:
with open(data_path / "story1.txt", 'r', encoding='unicode_escape') as f:
    corpus = f.read()

In [ ]:
f = open(r"D:\complete_content_new\Full-Stack-GenAI-Bootcamp-1.0\Class-06-07-11-12-Apr-Word2vec-with-practical\data\story1.txt",encoding='unicode_escape')

In [33]:
corpus

'Once upon a time, there was a young boy named Arjun who lived in a small town surrounded by hills and rivers. He was very curious about how things worked and always asked questions. His parents often smiled at his curiosity and encouraged him to learn more.\n\nEvery day after school, Arjun would sit with his books and explore different topics. He was especially fascinated by science and technology. He wanted to understand how machines think and whether they could become as intelligent as humans.\n\nOne day, while browsing through an old computer at school, Arjun discovered the concept of artificial intelligence. He read about how machines can learn from data, recognize patterns, and even make decisions. This idea completely amazed him.\n\nHe started imagining a future where machines could help humans solve complex problems. That day, he made a decision that he wanted to become an AI engineer.\n\nArjun began learning programming using free resources. At first, everything seemed difficu

In [34]:
sent_tokenize(corpus)

['Once upon a time, there was a young boy named Arjun who lived in a small town surrounded by hills and rivers.',
 'He was very curious about how things worked and always asked questions.',
 'His parents often smiled at his curiosity and encouraged him to learn more.',
 'Every day after school, Arjun would sit with his books and explore different topics.',
 'He was especially fascinated by science and technology.',
 'He wanted to understand how machines think and whether they could become as intelligent as humans.',
 'One day, while browsing through an old computer at school, Arjun discovered the concept of artificial intelligence.',
 'He read about how machines can learn from data, recognize patterns, and even make decisions.',
 'This idea completely amazed him.',
 'He started imagining a future where machines could help humans solve complex problems.',
 'That day, he made a decision that he wanted to become an AI engineer.',
 'Arjun began learning programming using free resources.',


In [35]:
raw_sent= sent_tokenize(corpus)

In [37]:
for i in raw_sent:
    print(simple_preprocess(i))

['once', 'upon', 'time', 'there', 'was', 'young', 'boy', 'named', 'arjun', 'who', 'lived', 'in', 'small', 'town', 'surrounded', 'by', 'hills', 'and', 'rivers']
['he', 'was', 'very', 'curious', 'about', 'how', 'things', 'worked', 'and', 'always', 'asked', 'questions']
['his', 'parents', 'often', 'smiled', 'at', 'his', 'curiosity', 'and', 'encouraged', 'him', 'to', 'learn', 'more']
['every', 'day', 'after', 'school', 'arjun', 'would', 'sit', 'with', 'his', 'books', 'and', 'explore', 'different', 'topics']
['he', 'was', 'especially', 'fascinated', 'by', 'science', 'and', 'technology']
['he', 'wanted', 'to', 'understand', 'how', 'machines', 'think', 'and', 'whether', 'they', 'could', 'become', 'as', 'intelligent', 'as', 'humans']
['one', 'day', 'while', 'browsing', 'through', 'an', 'old', 'computer', 'at', 'school', 'arjun', 'discovered', 'the', 'concept', 'of', 'artificial', 'intelligence']
['he', 'read', 'about', 'how', 'machines', 'can', 'learn', 'from', 'data', 'recognize', 'patterns',

In [44]:
folder_path = current_dir/"data"
story = []
for filename in os.listdir(folder_path):
    file_path = os.path.join(folder_path, filename)
    with open(file_path, 'r', encoding ='unicode_escape') as f:
        corpus = f.read()
    raw_sent = sent_tokenize(corpus)
    for sent in raw_sent:
        story.append(simple_preprocess(sent))    
   


In [45]:
story

[['once',
  'upon',
  'time',
  'there',
  'was',
  'young',
  'boy',
  'named',
  'arjun',
  'who',
  'lived',
  'in',
  'small',
  'town',
  'surrounded',
  'by',
  'hills',
  'and',
  'rivers'],
 ['he',
  'was',
  'very',
  'curious',
  'about',
  'how',
  'things',
  'worked',
  'and',
  'always',
  'asked',
  'questions'],
 ['his',
  'parents',
  'often',
  'smiled',
  'at',
  'his',
  'curiosity',
  'and',
  'encouraged',
  'him',
  'to',
  'learn',
  'more'],
 ['every',
  'day',
  'after',
  'school',
  'arjun',
  'would',
  'sit',
  'with',
  'his',
  'books',
  'and',
  'explore',
  'different',
  'topics'],
 ['he',
  'was',
  'especially',
  'fascinated',
  'by',
  'science',
  'and',
  'technology'],
 ['he',
  'wanted',
  'to',
  'understand',
  'how',
  'machines',
  'think',
  'and',
  'whether',
  'they',
  'could',
  'become',
  'as',
  'intelligent',
  'as',
  'humans'],
 ['one',
  'day',
  'while',
  'browsing',
  'through',
  'an',
  'old',
  'computer',
  'at',
  'sc

In [46]:
story[0]

['once',
 'upon',
 'time',
 'there',
 'was',
 'young',
 'boy',
 'named',
 'arjun',
 'who',
 'lived',
 'in',
 'small',
 'town',
 'surrounded',
 'by',
 'hills',
 'and',
 'rivers']

In [47]:
len(story)

44

In [48]:
story[0]

['once',
 'upon',
 'time',
 'there',
 'was',
 'young',
 'boy',
 'named',
 'arjun',
 'who',
 'lived',
 'in',
 'small',
 'town',
 'surrounded',
 'by',
 'hills',
 'and',
 'rivers']

In [49]:
story[1]

['he',
 'was',
 'very',
 'curious',
 'about',
 'how',
 'things',
 'worked',
 'and',
 'always',
 'asked',
 'questions']

In [51]:
custom_models = gensim.models.Word2Vec(story, vector_size=150, window=10, min_count=1)

In [52]:
custom_models

In [53]:
custom_models.build_vocab(story)

In [54]:
custom_models.corpus_count

44

In [55]:
custom_models.train(story, total_examples=custom_models.corpus_count, epochs=10)

(3392, 5220)

(3392, 5220) Part Meaning 3392 number of documents 5220 number of unique words

In [56]:
custom_models.wv["arjun"]

array([-0.00151581, -0.00421342,  0.00605878, -0.00663035,  0.00350448,
        0.00377938, -0.00119058,  0.00707785,  0.00601832, -0.0029052 ,
       -0.00373742, -0.00459522, -0.00614535, -0.00242868, -0.00450145,
       -0.00631547,  0.00094043,  0.00215407,  0.00406256, -0.00103225,
       -0.00354294,  0.00039096,  0.00152626,  0.00611874,  0.00018426,
        0.00423161, -0.00655704,  0.00444936, -0.00183366,  0.00511348,
       -0.00455955,  0.00143565, -0.00041108, -0.00023831, -0.00428121,
       -0.00588446, -0.00340444, -0.00480115,  0.00153196, -0.00020568,
       -0.00075928,  0.00481528, -0.00258614, -0.00141312, -0.0055643 ,
       -0.00689027,  0.00603142, -0.00427604, -0.00382809,  0.0031073 ,
        0.00429179, -0.00535715,  0.00059402, -0.00348999,  0.00064138,
       -0.00232278, -0.00210429,  0.00174472,  0.00584485,  0.00487289,
       -0.00734284, -0.00474751,  0.00501236,  0.00100381, -0.00285944,
        0.00022896,  0.00610954, -0.00242809, -0.00763496,  0.00

In [57]:
len(custom_models.wv["arjun"])

150

In [58]:
custom_models.wv["farmers"]

array([ 2.3416779e-03, -4.9130386e-03,  6.0363058e-03, -3.2153970e-03,
       -4.5480067e-03, -9.3988754e-04,  3.0667433e-03,  9.2603546e-04,
        2.4564720e-03,  6.8825670e-03, -2.5564958e-03,  5.9863338e-03,
        1.8564018e-04,  3.5295154e-03,  4.5037325e-03,  1.7350269e-03,
        6.1225607e-03, -5.9150909e-03, -7.9551432e-04,  3.9806846e-03,
        5.9812400e-03,  4.3811966e-03,  7.3130574e-04,  1.4921579e-04,
        5.5508218e-03,  2.2697940e-03,  2.9563436e-03, -4.4946033e-03,
       -3.9977030e-04, -6.3325404e-03, -3.9872251e-04, -5.7215644e-03,
       -1.3041844e-03,  5.8692875e-03, -5.5121547e-03,  5.5810739e-03,
       -3.7107891e-03,  4.1586030e-03,  3.0498598e-03, -4.8647211e-03,
       -4.9531274e-03, -3.5725329e-03, -6.3521112e-03, -3.7713021e-03,
        1.0072127e-03, -5.8164732e-03, -6.5759500e-03,  4.7708885e-03,
        2.8487353e-03, -2.6861166e-03,  2.7643850e-03, -1.7103793e-03,
        6.0753012e-03,  2.1378121e-03,  2.4504466e-03, -5.9942370e-03,
      

In [59]:
len(custom_models.wv["farmers"])

150

In [64]:
custom_models.wv["ai"]
len(custom_models.wv["ai"])

150

In [65]:
custom_models.wv.most_similar('arjun')

[('be', 0.26127907633781433),
 ('slowly', 0.24630282819271088),
 ('lack', 0.23023143410682678),
 ('most', 0.2302093356847763),
 ('whenever', 0.2211436927318573),
 ('patterns', 0.19873970746994019),
 ('so', 0.1958550214767456),
 ('teachers', 0.18988949060440063),
 ('him', 0.18940508365631104),
 ('interest', 0.1893124282360077)]

In [66]:
custom_models.wv.most_similar('ai')

[('lack', 0.24967938661575317),
 ('decided', 0.22155466675758362),
 ('used', 0.2207057774066925),
 ('imagining', 0.20021125674247742),
 ('proper', 0.1980542093515396),
 ('passed', 0.18372367322444916),
 ('slowly', 0.17975418269634247),
 ('understand', 0.15630404651165009),
 ('school', 0.15340426564216614),
 ('learned', 0.1487777680158615)]

In [67]:
custom_models.wv.most_similar('engineer')

[('data', 0.25355860590934753),
 ('named', 0.22001877427101135),
 ('farmers', 0.19320029020309448),
 ('education', 0.18713346123695374),
 ('kept', 0.186874121427536),
 ('create', 0.17003676295280457),
 ('many', 0.16424235701560974),
 ('encouraged', 0.1640772521495819),
 ('became', 0.1611989587545395),
 ('solve', 0.15225423872470856)]

In [68]:
custom_models.wv.most_similar("data")

[('realized', 0.25864478945732117),
 ('engineer', 0.2535586655139923),
 ('struggling', 0.25203195214271545),
 ('boy', 0.18801626563072205),
 ('single', 0.14922790229320526),
 ('think', 0.1473940759897232),
 ('kept', 0.1407105028629303),
 ('complex', 0.13979299366474152),
 ('frustrated', 0.13802938163280487),
 ('dedication', 0.13712474703788757)]

In [70]:
custom_models.wv.doesnt_match(["arjun", "engineer", "programming", "farmers"])

'arjun'

In [69]:
custom_models.wv.doesnt_match(["data", "model", "system", "river"])

'model'

In [71]:
custom_models.wv.doesnt_match(["weather", "crops", "farmers", "technology"])

'weather'

google-news-million of Artical
neural and llm -> data hungry

In [72]:
custom_models.wv.similarity("arjun", "engineer")

-0.09138947

Assignment for this week
add more data
train for more epoch
and check the accuracy of model
check the embedding dimesion also
inshort play with all the parameter